### PR TITLE
Add some functions to convert active cell iterators

### PR DESCRIPTION
**Pull request by jppelteret**
Let's provide some free functions to convert between active cell iterators. This should take care of [one of our FAQs](https://github.com/dealii/dealii/wiki/Frequently-Asked-Questions#can-i-convert-triangulation-cell-iterators-to-dofhandler-cell-iterators).

The naming is up for debate: I couldn't think of anything better, so please feel free to make some further recommendations!

### CODE REVIEW

File: include/deal.II/dofs/dof_handler.h
Comment (bangerth) [inline include/deal.II/dofs/dof_handler.h:1679]:
```suggestion
 * A function that converts a Triangulation active cell iterator to a
```
I would drop the "Triangulation" here because it is a DoFHandler iterator. It's not wrong, but slightly confusing.

**Review by bangerth — COMMENTED** [review]
Yes, nice suggestion.

**Comment by bangerth** [comment]
Alternatively, and maybe easier to find: How about a member function in `TriaAccessor` so that you can write
```
  Triangulation::active_cell_iterator cell;
  ...;
  DoFHandler::active_cell_iterator dof_cell = cell->convert_to_dof_handler_iterator(dof_handler);
```
?

File: include/deal.II/dofs/dof_handler.h
Comment (jppelteret) [reply include/deal.II/dofs/dof_handler.h:1695]: Yes, you are right. I'll add this in.

**Comment by jppelteret** [comment]
@bangerth, I followed through with your suggestion, adding the conversion function to `CellAccessor` (which allows one to check if the iterator `is_active()`). What do you think? I think that its nice to have a single implementation, but I must admit that I never find getting to this point in the documentation particularly easy!

I know that the state of the PR is now messy -- I'll clean it up tomorrow.

**Comment by tjhei** [comment]
Awesome! A member function might be easier to discover even if it is hidden somewhere in ``CellAccessor``.

File: include/deal.II/grid/tria_accessor.h
Comment (tjhei) [inline include/deal.II/grid/tria_accessor.h:3165]: You could also call it
```suggestion
  as_dof_handler_iterator(
```
(but the current choice is fine as well)

Comment (jppelteret) [reply include/deal.II/grid/tria_accessor.h]: Changed! Thanks for the recommendation.

**Comment by marcfehling** [comment]
I don't feel strongly about the following, but should this conversion function be a member of the `DoFHandler` class, and not the `CellAccessor` class? So, some template function like `DoFHandler::create_cell_iterator(AccessorType)`.

We had this debate for the `CellId::to_cell()` function, and made it a member function `Triangulation::create_cell_iterator()`. See also #11362, #11365, #14070.

With this I'm adding another opinion to the discussion. What do you think?

---

> I know we prefere verbose VS non-verbose, but how about `operator()(DoFHandler &dh)`?

I must say that I also like @luca-heltai's idea.

File: include/deal.II/grid/tria_accessor.h
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.h:3158]:
```suggestion
   * to an active cell iterator of another DoFHandler. The @p iterator must be
```

File: source/grid/tria_accessor.cc
Comment (bangerth) [inline source/grid/tria_accessor.cc:2127]:
```suggestion
  Assert(is_active(), ExcMessage("The current iterator points to an inactive cell. You cannot convert it to an iterator to an active cell."));
```

**Review by bangerth — COMMENTED** [review]
OK to merge if you address the two small issues. You probably have to reindent too then.

It would be interesting to know whether there are places in the tutorials that could be converted this way. But I don't know how to find out.

**Comment by bangerth** [comment]
Actually, step-32.cc lines 2221 and 2488, and step-51.cc in lines 701 and 1131 :-) 

It would be very nice if these places were converted as well (and maybe the iterators made `const` in the process).

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Sorry, not in favor of `operator()`. That's the call operator and the syntax gives no indication what exactly it does.

I could be convinced to make it a member function of `DoFHandler`, though I don't think the question is important enough to warrant an extended discussion :-)

The patch is ok with me as it currently stands, but I won't merge unless others feel the same way.

(And you'll have to address the following syntax error:
```
/jenkins/workspace/dealii_PR-14143/examples/step-32/step-32.cc:2221:131: error: expected ‘,’ or ‘;’ before ‘)’ token

     const typename DoFHandler<dim>::active_cell_iterator temperature_cell = cell->as_dof_handler_iterator(temperature_dof_handler));
```
)

File: examples/step-32/step-32.cc
Comment (bangerth) [inline examples/step-32/step-32.cc:2221]:
```suggestion
    const typename DoFHandler<dim>::active_cell_iterator temperature_cell = cell->as_dof_handler_iterator(temperature_dof_handler);
```

**Review by bangerth — APPROVED** [review]

File: examples/step-32/step-32.cc
Comment (jppelteret) [reply examples/step-32/step-32.cc]: Fixed. Thanks for pointing that out.

**Review by marcfehling — APPROVED** [review]

**Review by tjhei — APPROVED** [review]
I like it!

**Comment by bangerth** [comment]
We've got three approvals. The next person to approve might as well merge!

**Review by masterleinad — APPROVED** [review]

**Pull request merged**
